src/saga-dense.h

```diff
@@ -122,72 +124,97 @@ Saga(Penalty&               penalty,
      unsigned&              n_iter,
      std::vector<unsigned>& return_codes,
      std::vector<double>&   losses,
-     const bool             debug) noexcept
+     const bool             debug,
+     const bool             cyclic,
+     const unsigned         B) noexcept
 {
   using namespace std;
-
+    
   double wscale = 1.0;
-
+    
   double wscale_update = 1.0 - alpha*gamma;
-
+    
   penalty.setParameters(gamma, alpha, beta);
-
+    
   // Gradient vector and change in gradient vector
-  Eigen::ArrayXd g        = Eigen::ArrayXd::Zero(n_classes);
-  Eigen::ArrayXd g_change = Eigen::ArrayXd::Zero(n_classes);
-
-  Eigen::ArrayXd linear_predictor(n_classes);
-
+  Eigen::ArrayXXd g                = Eigen::ArrayXXd::Zero(n_classes, B);
+  Eigen::ArrayXXd g_change         = Eigen::ArrayXXd::Zero(n_classes, B);
+    
+  Eigen::ArrayXXd linear_predictor = Eigen::ArrayXXd::Zero(n_classes, B);
+    
+  // change for weight and average gradient 
+  Eigen::ArrayXXd step             = Eigen::ArrayXXd::Zero(n_classes, n_features);
+    
   // Setup functor for checking convergence
   ConvergenceCheck convergence_check{w, tol};
-
+    
+  // Setup selected sample matrix
+  Eigen::MatrixXd subx = Eigen::MatrixXd::Zero(n_features, B);
+    
+  // epoch size
+  const unsigned epoch = floor(n_samples/B);  
+    
+  // Setup index generator
+  Eigen::ArrayXXi index = Eigen::ArrayXXi::Zero(B, epoch);
+  Eigen::ArrayXi  s_ind = Eigen::ArrayXi::Zero(B);
+    
   // Outer loop
   unsigned it_outer = 0;
   bool converged = false;
   do {
+      
+    // Pull samples
+    index = Index(n_samples, B, cyclic);
+      
     // Inner loop
-    for (unsigned it_inner = 0; it_inner < n_samples; ++it_inner) {
-
-      // Pull a sample
-      unsigned s_ind = floor(R::runif(0.0, n_samples));
-
-      linear_predictor = (w.matrix() * x.col(s_ind)).array()*wscale + intercept;
-
+    for (unsigned it_inner = 0; it_inner < epoch; ++it_inner) {
+        
+      // Pull a epoch
+      s_ind = index.col(it_inner);
+        
+      // Select samples
+      subx = SelectCol(x, s_ind);
```

Comment:
This seems really expensive - do we actually need to make a whole new submatrix here? 

How does this compare efficiency wise to just building up the linear_predictor one term at a time in a loop?

Reply:
Hi mentor, I have tested that it is more efficient to have subx with colwise operation for mini batch, than without subx and setup linear_predictor by enumerating over selected samples. It seems that new method for directly working with sub matrix by specifying a set of indices in Eigen will be released: http://eigen.tuxfamily.org/bz/show_bug.cgi?id=329. I can change for that later.

Reply:
Ok - I've opened #30 to keep an eye on this.